src/golang/lib/engine/authenticate.go

```diff
@@ -42,11 +46,40 @@ func AuthenticateLambdaConfig(ctx context.Context, authConf auth.Config) error {
 		lambda_utils.SnowflakeConnectorType,
 	}
 
-	for _, functionType := range functionsToShip {
-		err := lambda_utils.CreateLambdaFunction(functionType, lambdaConf.RoleArn)
-		if err != nil {
-			return errors.Wrap(err, "Unable to Create Lambda Function")
+	errGroup, _ := errgroup.WithContext(ctx)
+
+	// Run authentication only once since the credentials will be memorized.
+	err = lambda_utils.AuthenticateDockerToECR()
+	if err != nil {
+		return errors.Wrap(err, "Unable to Create Lambda Function.")
+	}
+
+	// Pull images on a currency of "MaxConcurrentDownload" to parallelize while avoiding pull timeout.
+	for i := 0; i < len(functionsToShip); i += MaxConcurrentDownload {
+		for j := 0; j < MaxConcurrentDownload; j++ {
+			if j+i < len(functionsToShip) {
+				lambdaFunctionType := functionsToShip[j+i]
+				errGroup.Go(func() error {
+					return lambda_utils.PullImageFromECR(lambdaFunctionType)
+				})
+			}
+		}
+		if err := errGroup.Wait(); err != nil {
+			return errors.Wrap(err, "Unable to Create Lambda Function.")
```

Comment:
Can we make these error messages more specific? Lets note that the failure happened at the pulling stage at least.